2000 North Carolina Congressional Districts

brianlzhou opened the pull request:
# 2000 North Carolina Congressional Districts

## Redistricting requirements
In North Carolina, according to [NCSL Redistricting Law 2000](https://web.archive.org/web/20041216185957/https://www.senate.mn/departments/scr/redist/red2000/Tab5appx.htm), districts must:

1. be contiguous
1. have equal populations
1. be geographically compact
1. preserve county and municipality boundaries as much as possible
1. follow the Voting Rights Act
1. change as few districts as possible
1. keep the delegation's current partisan balance
1. keep incumbents in separate districts
1. preserve cores of existing districts


### Algorithmic Constraints
We enforce a maximum population deviation of 0.5%.

## Data Sources
Data for North Carolina comes from the [ALARM Project's update](https://dataverse.harvard.edu/dataset.xhtml?persistentId=doi:10.7910/DVN/ZV5KF3) to [The Record of American Democracy](https://road.hmdc.harvard.edu/).

## Pre-processing Notes
No manual pre-processing decisions were necessary.

## Simulation Notes
We sample 10,000 districting plans for North Carolina across 5 independent runs of the SMC algorithm.
We then thinned the number of samples to 5,000. 
No special techniques were needed to produce the sample.


## Validation

<img width="3000" height="5400" alt="image" src="https://github.com/user-attachments/assets/338a7e57-3b13-4688-bd32-ae4e7cc2de46" />

<img width="1762" height="1072" alt="image" src="https://github.com/user-attachments/assets/e5ad86c6-794c-4033-81af-66ca20b1e9c8" />

<img width="1762" height="1072" alt="image" src="https://github.com/user-attachments/assets/9dfcaf53-abc1-4c79-9eb9-080749f1a75a" />


```
SMC with Merge-Split MCMC Steps: 6,000 sampled plans of 13 districts on 2,647 units
Plans sampled on Spanning Forest Space using the Uniform Valid Edge forward kernel.
SMC Parameters:
• `weight_type` = optimal
• `seq_alpha` = 1
• `pop_temper` = 0
Mergesplit Parameters:
• `total_ms_steps` = 12
• `mh_accept_per_smc` = 50
• `pair_rule` = uniform
Plan diversity 80% range: 0.6 to 0.8
Largest R-hat values for ordered summary statistics:
    comp_bbox_reock           comp_edge         comp_polsby       county_splits         muni_splits 
              1.008               1.000               1.005               1.001               1.003 
            ndshare                 ndv                 nrv            plan_dev            pop_aian 
              1.015               1.018               1.018               1.012               1.021 
          pop_asian           pop_black            pop_hisp            pop_nhpi           pop_other 
              1.018               1.033               1.010               1.012               1.013 
        pop_overlap             pop_two           pop_white total_county_splits   total_muni_splits 
              1.004               1.022               1.019               1.000               1.004 
          total_vap            vap_aian           vap_asian           vap_black            vap_hisp 
              1.020               1.022               1.020               1.029               1.014 
           vap_nhpi           vap_other             vap_two           vap_white 
              1.009               1.014               1.014               1.020 
• R-hat ≤ 1.05: 377
• 1.05 < R-hat ≤ 1.1: 0
• R-hat > 1.1: 0
• Total R-hats: 377
Sampling diagnostics for SMC run 1 of 6 (6,000 samples)
         Eff. samples (%) Acc. rate Log wgt. sd  Max. unique 
Split 1       997 (33.2%)      1.9%        1.34 1,869 ( 99%) 
Split 2     1,155 (38.5%)     95.5%        1.19 1,349 ( 71%) 
Split 3     1,340 (44.7%)     71.0%        0.96 1,366 ( 72%) 
Split 4     1,914 (63.8%)      1.2%        0.76   708 ( 37%) 
Split 5     1,464 (48.8%)     52.5%        0.84 1,600 ( 84%) 
Split 6     1,700 (56.7%)     40.4%        0.74 1,518 ( 80%) 
Split 7     1,933 (64.4%)     31.0%        0.66 1,538 ( 81%) 
Split 8     2,165 (72.2%)     25.4%        0.58 1,583 ( 83%) 
Split 9     2,286 (76.2%)     20.3%        0.53 1,579 ( 83%) 
Split 10    2,390 (79.7%)     16.7%        0.48 1,615 ( 85%) 
Split 11    2,523 (84.1%)     13.9%        0.41 1,625 ( 86%) 
Split 12    2,742 (91.4%)     11.0%        0.30 1,596 ( 84%) 
Resample    2,742 (91.4%)       NA%          NA 2,666 (141%) 

Sampling diagnostics for Mergesplit Steps of SMC run 1 of 6 (6,000 samples)
           Acc. rate MS Steps per Plan
MS Step 1       0.8%              2600
MS Step 2      42.2%              6000
MS Step 3      34.0%               150
MS Step 4      18.4%               150
MS Step 5      28.9%               300
MS Step 6      27.3%               200
MS Step 7      26.4%               200
MS Step 8      26.6%               200
MS Step 9      27.5%               200
MS Step 10     28.5%               200
MS Step 11     29.9%               200
MS Step 12     31.3%               200

•  Watch out for low effective samples, very low acceptance rates (less than 1%), large std. devs. of
the log weights (more than 3 or so), and low numbers of unique plans. R-hat values for summary
statistics should be between 1 and 1.05.
```

```
   bvap_rank     dem
       <dbl>   <dbl>
 1         1 0      
 2         2 0      
 3         3 0      
 4         4 0.0015 
 5         5 0.0045 
 6         6 0.00383
 7         7 0.0425 
 8         8 0.0702 
 9         9 0.142  
10        10 0.176  
11        11 0.491  
12        12 0.920  
13        13 0.964  
```

```
  n_black_perf    n
1            2 5755
2            3  245
```

## Checklist

- [x] I have followed the [instructions](https://github.com/alarm-redist/fifty-states/blob/main/CONTRIBUTING.md)
- [x] I have updated the [tracker](https://docs.google.com/spreadsheets/d/1k_tYLoE49W_DCK1tcWbouoYZFI9WD76oayEt5TOmJg4/edit#gid=453387933)
- [x] All `TODO` lines from the template code have been removed
- [x] I have merged in the main branch and then recalculated summary statistics
- [x] I have run `enforce_style()` to format my code
- [x] The documentation copied above is up-to-date 
- [x] There are no data files in this pull request
- [x] None of the file output paths (for the `redist_map` and `redist_plans` objects, and summary statistics) have been edited

@tylersimko

brianlzhou commented:
It seems like there are some N/A values in NRV/NDV which is preventing me from creating an accurate dem performance plot with the logit regression - I'm running into the same issue as with AL
`Error in uniroot(function(shift) { : f.lower = f(lower) is NA
5.
stop("f.lower = f(lower) is NA")
4.
res <- uniroot(function(shift) {
stats::weighted.mean(plogis(ldvs + shift), turn) - target
}, c(-1, 1), tol = tol)
3.
x |>
logit_shift_baseline(ndv = ndv, nrv = nrv, target = target)
2.
FUN(X[[i]], ...)
1.
lapply(group_split(group_by(nc_shp, county_fips)), function(x) {
meds <- filter(medsl_cty, county == x$county_fips[1])
target <- meds$dshare_00[1]
logit_shift_baseline(x, ndv = ndv, nrv = nrv, target = target) ...`
@christopherkenny I believe you did the ROAD data for NC as an example, would you be able to check this out?

Including this image of the enacted plan for NC's three MMDs - this is impossible to recreate in simulation save for loosening every constraint by a wide margin
<img width="2560" height="999" alt="image" src="https://github.com/user-attachments/assets/5c87e558-b601-4ba6-ab4b-29ccc41b6cda" />

tylersimko reviewed (changes requested):
Just putting comment above in the review format: 

> @brianlzhou I know you can't make the meeting today, so no worries. Were you able to fix the election data issue above and make a performance plot?

brianlzhou commented:
@tylersimko Just updated the PR with the latest including validation plots.

tylersimko commented:
Thanks @brianlzhou ! This looks good. Once you've had a chance to implement the changes from the meeting today, please just ping me again and looks like this will be good to merge in.

brianlzhou commented:
updated - looks like it's good to go! 

## Validation

<img width="3000" height="4500" alt="image" src="https://github.com/user-attachments/assets/d44b5c1d-82ab-4c80-9550-dade974a783d" />

<img width="1762" height="1060" alt="image" src="https://github.com/user-attachments/assets/cdf23005-4e98-42e0-9069-50463b43ecc8" />


```
SMC with Merge-Split MCMC Steps: 6,000 sampled plans of 13 districts on 2,647 units
Plans sampled on Spanning Forest Space using the Uniform Valid Edge forward kernel.
SMC Parameters:
• `weight_type` = optimal
• `seq_alpha` = 1
• `pop_temper` = 0
Mergesplit Parameters:
• `total_ms_steps` = 12
• `mh_accept_per_smc` = 30
• `pair_rule` = uniform
Plan diversity 80% range: 0.59 to 0.80
Largest R-hat values for ordered summary statistics:
    comp_edge   comp_polsby county_splits   muni_splits       ndshare           ndv           nrv 
        1.012         1.012         1.001         1.014        ❌1.06         1.038       ❌1.056 
     plan_dev      pop_aian     pop_asian     pop_black      pop_hisp      pop_nhpi     pop_other 
        1.006         1.037         1.033         1.014         1.019         1.019         1.023 
  pop_overlap       pop_two     pop_white     total_vap      vap_aian     vap_asian     vap_black 
        1.029         1.046         1.045         1.028         1.039         1.029         1.018 
     vap_hisp      vap_nhpi     vap_other       vap_two     vap_white 
        1.016         1.021         1.023         1.048         1.040 
• R-hat ≤ 1.05: 336
• 1.05 < R-hat ≤ 1.1: 2
• R-hat > 1.1: 0
• Total R-hats: 338
ℹ ALERT: Chains have weakly converged.
Sampling diagnostics for SMC run 1 of 6 (6,000 samples)
         Eff. samples (%) Acc. rate Log wgt. sd  Max. unique 
Split 1       729 (36.4%)      2.0%        1.30 1,258 (100%) 
Split 2       772 (38.6%)     96.0%        1.18   897 ( 71%) 
Split 3       904 (45.2%)     72.2%        0.96   949 ( 75%) 
Split 4     1,337 (66.8%)      1.1%        0.75   463 ( 37%) 
Split 5       962 (48.1%)     51.7%        0.85 1,103 ( 87%) 
Split 6     1,142 (57.1%)     39.9%        0.74   998 ( 79%) 
Split 7     1,288 (64.4%)     30.7%        0.65 1,043 ( 83%) 
Split 8     1,441 (72.0%)     25.3%        0.59 1,049 ( 83%) 
Split 9     1,510 (75.5%)     21.5%        0.53 1,055 ( 83%) 
Split 10    1,614 (80.7%)     17.2%        0.46 1,053 ( 83%) 
Split 11    1,711 (85.5%)     13.4%        0.40 1,106 ( 87%) 
Split 12    1,847 (92.3%)     11.0%        0.29 1,069 ( 85%) 
Resample    1,847 (92.3%)       NA%          NA 1,778 (141%) 

Sampling diagnostics for Mergesplit Steps of SMC run 1 of 6 (6,000 samples)
           Acc. rate MS Steps per Plan
MS Step 1       0.9%              1500
MS Step 2      42.5%              3300
MS Step 3      34.4%                90
MS Step 4      18.8%                90
MS Step 5      29.7%               180
MS Step 6      28.1%               120
MS Step 7      27.5%               120
MS Step 8      27.5%               120
MS Step 9      27.9%               120
MS Step 10     28.8%               120
MS Step 11     30.3%               120
MS Step 12     31.4%               120
```

```
   bvap_rank      dem
       <dbl>    <dbl>
 1         1 0       
 2         2 0       
 3         3 0       
 4         4 0       
 5         5 0.000167
 6         6 0.0015  
 7         7 0.0282  
 8         8 0.047   
 9         9 0.0495  
10        10 0.0777  
11        11 0.360   
12        12 0.925   
13        13 0.975   
```

```
  n_black_perf    n
1            2 5776
2            3  224
```
